### PR TITLE
feat: mutex-manager

### PR DESCRIPTION
**Pull request by icezohu**
with async-mutex by default
with redis-mutex when REDIS_MUTEX is on
with other mutex when registered and called with relative option 
test: mutex-manager.test with sequence field and sort field

by Jacob

> **Note**
> This is a template for submitting a new feature. Use the bug fix template if you're submitting a bug fix pull request by adding `template=bug_fix.md` to your pull request URL.

# Description (需求描述)
Describe the new feature or modification to an existing feature clearly and consciously.

# Motivation (需求背景)
Explain the reason for adding or modifying this feature.

# Key changes (关键改动）
Provide a technically detailed description of the key changes made.
- Frontend (前端)
- Backend (后端)

# Test plan (测试计划)
## Suggestions (测试建议)
Provide any suggestions or recommendations for improvements in the testing plan.

## Underlying risk (潜在风险)
Identify any potential risks or issues that may arise from the new feature or modification.

# Showcase (结果展示）
Including any screenshots of the new feature or modification.

### CODE REVIEW

File: packages/core/utils/src/mutex-manager/mutex-manager.ts [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  public static getInstance(): MutexManager {
+    if (!MutexManager.instance) {
+      MutexManager.instance = new MutexManager();
+    }
+    return MutexManager.instance;
+  }
+
+  private constructor() {}
+
+  public setupEventListeners(app: any) {
```
Comment (chenos) [inline packages/core/utils/src/mutex-manager/mutex-manager.ts:45]: 这个设计很奇怪，理想情况不应该依赖 App 相关模块

File: packages/core/server/package.json [hunk 2]
```diff
@@ -25,7 +25,10 @@
     "lodash": "^4.17.21",
     "moment": "^2.29.1",
     "semver": "^7.3.7",
-    "xpipe": "^1.0.5"
+    "xpipe": "^1.0.5",
+    "ioredis": "^5.3.2",
+    "redlock": "^5.0.0-beta.2",
+    "async-mutex": "^0.3.2"
```
Comment (chenos) [inline packages/core/server/package.json:31]: 这些依赖的声明也不适合放在这里

**Review by chenos — CHANGES_REQUESTED** [review]

File: packages/core/utils/src/mutex-manager/mutex-manager.ts (hunk 1 above)
Comment (icezohu) [reply packages/core/utils/src/mutex-manager/mutex-manager.ts:45]: 这个是为了单元测试的时候，能够断开redis的连接，不然测试任务结束不了。我再看看有没有别的办法。

File: packages/core/database/src/fields/sort-field.ts [hunk 3]
```diff
@@ -27,11 +25,27 @@ export class SortField extends Field {
       }
     }
 
-    await sortFieldMutex.runExclusive(async () => {
-      const max = await model.max<number, any>(name, { ...options, where });
-      const newValue = (max || 0) + 1;
-      instance.set(name, newValue);
+    const mutexName = MutexManager.nameHelper({
+      db_dialect: this.database.options.dialect,
```
[the remark this answers is not in the document]
Comment (icezohu) [reply packages/core/database/src/fields/sort-field.ts]: 已修正

File: packages/core/utils/src/index.ts [hunk 4]
```diff
@@ -14,3 +14,4 @@ export * from './requireModule';
 export * from './toposort';
 export * from './uid';
 export * from './json-templates';
+export * from './mutex-manager';
```
[the remark this answers is not in the document]
Comment (icezohu) [reply packages/core/utils/src/index.ts]: 已重构

File: packages/core/utils/src/mutex-manager/mutex-manager.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+
+export interface MutexNameHelperConfig {
+  db_dialect?: string;
+  db_host?: string;
+  db_port?: number;
+  db_storage?: string;
+  db_name?: string;
+  table_name?: string;
+  other?: string;
+  action?: string;
```
[the remark this answers is not in the document]
Comment (icezohu) [reply packages/core/utils/src/mutex-manager/mutex-manager.ts]: 已修正

File: packages/core/server/package.json (hunk 2 above)
Comment (icezohu) [reply packages/core/server/package.json]: 已修正

File: packages/core/utils/src/mutex-manager/mutex-manager.ts (hunk 1 above)
Comment (icezohu) [reply packages/core/utils/src/mutex-manager/mutex-manager.ts]: 删除了相关代码之后，有9个测试用例出错，大部分报 `ExecutionError: The operation was unable to achieve a quorum during its retry window.`

**Pull request closed without merging**